### DISCUSSION TITLE
How to track user attribution for changes in slate-yjs

### DISCUSSION THREAD

**Discussion by KentonParton** [BitPhinix/slate-yjs#427] [Q&A]
Hey there! 👋 

I'm working on a collaborative editing project using slate-yjs with the [Plate editor](https://platejs.org/), and I could use some guidance on tracking user contributions.

**What I'm trying to do:**
I need to keep track of which user made what changes in the document - basically knowing who wrote or edited each piece of content. This would be super helpful for comparing versions and seeing who made specific changes.

**Current setup:**
- Using slate-yjs with the [Plate](https://github.com/udecode/plate) editor for the frontend
- Hocuspocus for the backend

**The tricky part:**
I'm not quite sure how to:
1. Associate user info with each change
2. Store this info so it sticks around in the document

**What else:**
1. We would also need to verify the userId of each change submitted to the server matches the userId on the server. I think we could do this in the [before-handle-message](https://tiptap.dev/docs/hocuspocus/server/hooks#before-handle-message)

Any tips or pointers would be really appreciated!

Thanks in advance! 🙏

**Comment by BitPhinix** [comment]
It would need custom handling, although this shouldn't be too hard - yjs keeps track of which client id caused a specific change as part of its conflict resolution, so it already has "attribution". So you would just need to store which client id belongs to which user (and do some custom handling to ensure updates for each corresponding client id are only submitted by the associated userids).

I did this for someone before but forgot the implementation details - I'm kind of out of the yjs world atm.

**Reply by KentonParton** [reply]
Is the client ID in the YJS Doc?

Is this somewhat related to withYHistory? Something I don't quite understand is whether withYHistory is in memory or persisted to the YJS doc. I'm trying to add support for document versioning as users interact with it. The easy but naive approach is duplicate the whole Y Doc but curious if withYHistory already stores the history of changes. Perhaps there is an easy way to display points in time?

**Reply by KentonParton** [reply]
After reading this thread https://discuss.yjs.dev/t/how-to-recover-to-the-specified-version/2301/7 something like the below might work. Create and store the snapshot in DB. Then to restore call revertToSnapshot and somehow push the new Y Doc in YJS Plugin.

```typescript
"use client";

import React, { useEffect, useState } from 'react';
import * as Y from 'yjs';

interface Version {
  id: string;
  timestamp: number;
  snapshot: Uint8Array;
  description: string;
}

// Proper revert function based on Himself65's approach
const revertToSnapshot = (doc: Y.Doc, snapshot: Uint8Array) => {
  // Temporarily disable garbage collection
  doc.gc = false;
  
  // Create a temporary doc from the snapshot
  const snapshotDoc = new Y.Doc();
  Y.applyUpdate(snapshotDoc, snapshot);
  
  // Get state vectors
  const currentStateVector = Y.encodeStateVector(doc);
  const snapshotStateVector = Y.encodeStateVector(snapshotDoc);
  
  // Encode changes since snapshot
  const changesSinceSnapshot = Y.encodeStateAsUpdate(doc, snapshotStateVector);
  
  // Create UndoManager for the snapshot doc
  const undoManager = new Y.UndoManager([snapshotDoc.getText('content')]);
  
  // Apply changes and undo them to create revert operation
  Y.applyUpdate(snapshotDoc, changesSinceSnapshot);
  undoManager.undo();
  
  // Create and apply the revert update
  const revertUpdate = Y.encodeStateAsUpdate(snapshotDoc, currentStateVector);
  Y.applyUpdate(doc, revertUpdate, { origin: 'revert' });
  
  // Re-enable garbage collection
  doc.gc = true;
};

const BetterVersioning = () => {
  const [doc] = useState(() => new Y.Doc());
  const [text] = useState(() => doc.getText('content'));
  const [value, setValue] = useState('');
  const [versions, setVersions] = useState<Version[]>([]);
  const [diffView, setDiffView] = useState({ from: '', to: '' });

  useEffect(() => {
    text.observe(() => {
      setValue(text.toString());
    });

    // Create initial version
    createVersion('Initial version');
  }, []);

  const createVersion = (description: string) => {
    const version: Version = {
      id: crypto.randomUUID(),
      timestamp: Date.now(),
      snapshot: Y.encodeStateAsUpdate(doc),
      description
    };
    
    setVersions(prev => [...prev, version]);
  };

  // Updated restore function using proper revert
  const restoreVersion = (version: Version) => {
    revertToSnapshot(doc, version.snapshot);
  };

  const compareVersions = (v1: Version, v2: Version) => {
    const doc1 = new Y.Doc();
    const doc2 = new Y.Doc();
    
    Y.applyUpdate(doc1, v1.snapshot);
    Y.applyUpdate(doc2, v2.snapshot);
    
    setDiffView({
      from: doc1.getText('content').toString(),
      to: doc2.getText('content').toString()
    });
  };

  const handleChange = (e: React.ChangeEvent<HTMLTextAreaElement>) => {
    const newValue = e.target.value;
    doc.transact(() => {
      text.delete(0, text.length);
      text.insert(0, newValue);
    });
  };

  return (
    <div className="p-4 max-w-4xl mx-auto">
      <div className="space-y-4">
        <div>
          <h2 className="text-lg font-semibold mb-2">Editor</h2>
          <textarea
            value={value}
            onChange={handleChange}
            className="w-full h-32 p-2 border rounded"
          />
        </div>

        <div>
          <button
            onClick={() => createVersion(`Version at ${new Date().toLocaleTimeString()}`)}
            className="px-4 py-2 bg-blue-500 text-white rounded hover:bg-blue-600"
          >
            Create Version
          </button>
        </div>

        <div>
          <h2 className="text-lg font-semibold mb-2">Versions</h2>
          <div className="space-y-2">
            {versions.map((version, index) => (
              <div key={version.id} className="p-2 border rounded flex justify-between items-center">
                <div>
                  <div className="font-medium">Version {index + 1}</div>
                  <div className="text-sm text-gray-600">
                    {new Date(version.timestamp).toLocaleString()}
                  </div>
                  <div className="text-sm">{version.description}</div>
                </div>
                <div className="space-x-2">
                  <button
                    onClick={() => restoreVersion(version)}
                    className="px-3 py-1 bg-gray-200 rounded hover:bg-gray-300"
                  >
                    Restore
                  </button>
                  {index > 0 && (
                    <button
                      onClick={() => compareVersions(versions[index - 1], version)}
                      className="px-3 py-1 bg-gray-200 rounded hover:bg-gray-300"
                    >
                      Compare with Previous
                    </button>
                  )}
                </div>
              </div>
            ))}
          </div>
        </div>

        {diffView.from !== '' && diffView.to !== '' && (
          <div className="mt-4">
            <h2 className="text-lg font-semibold mb-2">Diff View</h2>
            <div className="grid grid-cols-2 gap-4">
              <div>
                <h3 className="font-medium mb-1">Previous Version</h3>
                <pre className="p-2 bg-gray-100 rounded whitespace-pre-wrap">
                  {diffView.from}
                </pre>
              </div>
              <div>
                <h3 className="font-medium mb-1">Current Version</h3>
                <pre className="p-2 bg-gray-100 rounded whitespace-pre-wrap">
                  {diffView.to}
                </pre>
              </div>
            </div>
          </div>
        )}
      </div>
    </div>
  );
};

export default BetterVersioning;
```

**Reply by KentonParton** [reply]
Hi @BitPhinix ,

Does this approach seem like the right direction for managing document versioning and reverting changes?

Also, is there a way to manually push the changes created when a snapshot is applied back into the Y.Doc’s history through the YJS plugin? If so, how would I go about doing that?

Thanks for your help!

**Reply by BitPhinix** [reply]
The structure seems ok. You might want to consider using a `Y.snapshot` if you are turning off gc anyway - although saving full updates like you are doing now might enable you to re-enable GC. So both work, doesn't really matter.

If you want to visualize changes between snapshots you can use [`.toDelta(snapshot, prevSnapshot)`](https://github.com/yjs/yjs/blob/0419b74315769a4867fd9655fab8f7f0fdccda3c/src/types/YText.js#L1009) which will give you the changes between two snapshots which you then can visualize (you have to do this recursively over all elements in the yjs doc).

As to applying changes from history versions to the latest version of the document - that should just happen automatically when you get the state of the modified snapshot and apply that to the latest document state. In the end a snapshot is just a set of ids of "characters" (oversimplifying here ofc) that should be present / deleted from the document. Since the latest state has all changes that where in the snapshot you started from originally applying the update of the modified snapshot should just apply these changes.

**Comment by trannhutle** [comment]
@KentonParton, I am working on the change tracking feature like google and notion. are you trying to solve the same problem. If you are solving the same issue. Can you please show me what is a direction for that?
Thank you very much!

**Reply by KentonParton** [reply]
Hey @trannhutle yes, it would be great to have user change tracking. Unfortunately, it's not something I'm familiar enough with yet to dive into unless more guidance is given.

If you end up giving it a go, please share a link in this thread.

**Reply by trannhutle** [reply]
I have been researching from the last couple of weeks. There is not much progress  and I am stuck at the solution. 
I would like to reach out to our community for ideas sharing.
